eth/downloader/stagedsync_stage_hashcheck.go

```diff
@@ -78,3 +98,238 @@ func (d *Downloader) unwindHashCheckStage(unwindPoint uint64) error {
+	})
+	if err != nil {
+		return err
+	}
+
+	sort.Sort(buffer)
+	var file string
+	file, err = buffer.FlushToDisk(datadir)
+	if err != nil {
+		return err
+	}
+	files = append(files, file)
+	return mergeTempFilesIntoBucket(db, files, destBucket)
+}
+
+func transformPlainStateKey(key []byte) ([]byte, error) {
+	switch len(key) {
+	case common.AddressLength:
+		// account
+		hash, err := common.HashData(key)
+		return hash[:], err
+	case common.AddressLength + common.IncarnationLength + common.HashLength:
+		// storage
+		address, incarnation, key := dbutils.PlainParseCompositeStorageKey(key)
+		addrHash, err := common.HashData(address[:])
+		if err != nil {
+			return nil, err
+		}
+		secretKey, err := common.HashData(key[:])
+		if err != nil {
+			return nil, err
+		}
+		compositeKey := dbutils.GenerateCompositeStorageKey(addrHash, incarnation, secretKey)
+		return compositeKey, nil
+	default:
+		// no other keys are supported
+		return nil, fmt.Errorf("could not convert key from plain to hashed, unexpected len: %d", len(key))
+	}
+}
+
+func transformContractCodeKey(key []byte) ([]byte, error) {
+	if len(key) != common.AddressLength+common.IncarnationLength {
+		return nil, fmt.Errorf("could not convert code key from plain to hashed, unexpected len: %d", len(key))
+	}
+	address, incarnation := dbutils.PlainParseStoragePrefix(key)
+
+	addrHash, err := common.HashData(address[:])
+	if err != nil {
+		return nil, err
+	}
+
+	compositeKey := dbutils.GenerateStoragePrefix(addrHash[:], incarnation)
+	return compositeKey, nil
+}
+
+type sortableBufferEntry struct {
+	key   []byte
+	value []byte
+}
+
+type sortableBuffer struct {
+	entries     []sortableBufferEntry
+	size        int
+	OptimalSize int
+	encoder     *codec.Encoder
+}
+
+func (b *sortableBuffer) Put(k, v []byte) {
+	b.size += len(k)
+	b.size += len(v)
+	b.entries = append(b.entries, sortableBufferEntry{k, v})
+}
+
+func (b *sortableBuffer) Size() int {
+	return b.size
+}
+
+func (b *sortableBuffer) Len() int {
+	return len(b.entries)
+}
+
+func (b *sortableBuffer) Less(i, j int) bool {
+	return bytes.Compare(b.entries[i].key, b.entries[j].key) < 0
+}
+
+func (b *sortableBuffer) Swap(i, j int) {
+	b.entries[i], b.entries[j] = b.entries[j], b.entries[i]
+}
+
+func (b *sortableBuffer) FlushToDisk(datadir string) (string, error) {
+	bufferFile, err := ioutil.TempFile(datadir, "tg-sync-sortable-buf")
+	if err != nil {
+		return "", err
+	}
+	defer bufferFile.Close() //nolint:errcheck
+
+	filename := bufferFile.Name()
+	w := bufio.NewWriter(bufferFile)
+	defer w.Flush() //nolint:errcheck
+	b.encoder.Reset(w)
+
+	for i := range b.entries {
+		err = writeToDisk(b.encoder, b.entries[i].key, b.entries[i].value)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	b.entries = b.entries[:0] // keep the capacity
+	return filename, nil
+}
+
+func newSortableBuffer() *sortableBuffer {
+	return &sortableBuffer{
+		entries:     make([]sortableBufferEntry, 0),
+		size:        0,
+		OptimalSize: 1024 * 1024,
```

Comment:
Way too small for the optimal size, if it is in bytes. Make it 256 Mb, this is what I used in the stage 6 and stage 7

Reply:
makes sense